src/box/relay.cc

```diff
@@ -832,6 +847,52 @@ relay_thread_on_stop(void *arg)
 	relay->tx.is_raft_enabled = false;
 }
 
+/** The trigger_vclock_sync call message. */
+struct relay_trigger_vclock_sync_msg {
+	/** Parent cbus message. */
+	struct cbus_call_msg base;
+	/** The queried relay. */
+	struct relay *relay;
+	/** Sync value returned from relay. */
+	uint64_t vclock_sync;
+};
+
+/** A callback to free the message once it returns to tx thread. */
+static int
+relay_trigger_vclock_sync_msg_free(struct cbus_call_msg *msg)
+{
+	free(msg);
+	return 0;
+}
+
+/** Relay side of the trigger_vclock_sync call. */
+static int
+relay_trigger_vclock_sync_f(struct cbus_call_msg *msg)
+{
+	struct relay_trigger_vclock_sync_msg *m =
+		(struct relay_trigger_vclock_sync_msg *)msg;
+	m->vclock_sync = m->relay->sent_vclock_sync + 1;
+	m->relay->need_new_vclock_sync = true;
+	return 0;
+}
+
+int
+relay_trigger_vclock_sync(struct relay *relay, uint64_t *vclock_sync,
+			  double deadline)
+{
+	struct relay_trigger_vclock_sync_msg *msg =
+		(struct relay_trigger_vclock_sync_msg *)xmalloc(sizeof(*msg));
+	msg->relay = relay;
+	double timeout = deadline - ev_monotonic_now(loop());
+	if (cbus_call_timeout(&relay->relay_pipe, &relay->tx_pipe, &msg->base,
+			      relay_trigger_vclock_sync_f,
+			      relay_trigger_vclock_sync_msg_free, timeout) < 0)
```

Comment:
I don't think we need a timeout here - triggering a sync is a fast operation that isn't supposed to stall. Having a timeout here only looks confusing IMO.

Reply:
There are no guarantees that relay isn't stalled outside of `cbus_proces` loop forever. For example, it might be sending a WAL, and that WAL might be appended at a high rate.

Besides, there's no way to cancel a `cbus_call` without a timeout.

I'd rather stick with the timeout. Besides, it's TIMEOUT_INFINITY, when user doesn't specify anything else.

Reply:
OK. Let's leave a timeout - not critical. We can remove it in future, if we make the tx <-> relay bus more reliable.